tools/copy-rom/copy-rom.8
.\" $Id: copy-rom.8 181 2005-05-07 06:51:15Z pdh $
.\"
.\" Copyright (C) 2004, 2005  Martin Michlmayr <tbm@cyrius.com>
.\" This manual is freely distributable under the terms of the GPL.
.\" It was originally written for Debian GNU/Linux (but may be used
.\" by others).
.\"

.TH COPY\-ROM 8 "May 2005"

.SH NAME
copy\-rom \- output the MD5 checksum of a Cobalt's ROM, and optionally
save a copy of the ROM to a file.

.SH SYNOPSIS
.B copy\-rom
[\fIfile\fR]

.SH DESCRIPTION
.PP
.B copy\-rom
calculates the MD5 checksum of the data stored in a Cobalt's ROM and
outputs the checksum. If a filename is given a copy of the ROM
is saved to the file.

.SH SEE ALSO
.PP
.BR md5sum (1)

.SH FILES
.PP
.IR /usr/share/doc/colo/COBALT.MD5
- MD5 checksums of original Cobalt firmware

.SH AUTHOR
.PP
Peter Horton <pdh@colonel\-panic.org>



CHANGES
# $Id: CHANGES 195 2006-08-10 22:03:35Z pdh $

CoLo 1.22 (2006-08-10)
----------------------

*	Fix build failure on GCC 4.1.

*	Tweaked INSTALL.
	[Martin Michlmayr]

*	Fixed 'netcon' to not cause an assert in UDP code.

*	Added support for CF cards in CF/IDE adapters (not tested).

CoLo 1.21 (2006-01-22)
----------------------

*	Merged "screen" and "minicom" keymaps.

*	Added "delete word" to line editor ('<ctrl>w').

*	Added 'relocate' command which relocates the initrd image to follow the
	kernel image. Required to use initrd images with current 2.6.x kernels.

*	'mount' command can now be aborted.

CoLo 1.20 (2006-01-16)
----------------------

*	Fix menu cancel/abort/error returns.

*	Update NV format to include keymap.

*	Added keymap for "screen".

*	Wait for network interface transmit ring to drain when disabling network
	interface. This should improve net console output.

*	Added header file dependency checking to stage1, stage2 and chain.

CoLo 1.19 (2005-04-03)
----------------------

*	Fixed stupid bug in ordering of LCD initialisation. This would cause
	installs of 1.17 or 1.18 which weren't an upgrade from 1.16 to crash
	early in stage2.

*	Added "README.tools" to tools directory.

*	Renamed 'md5rom' to 'copy-rom' and added the ability to save a copy of the
	ROM to a file.

*	Moved PCI timeout / retry counter initialisation to PCI initialisation
	code. We now use a retry count of 255 rather than retrying forever. The
	kernel needs a similiar fix.

*	Improved PCI bus scan.

CoLo 1.18 (2005-03-22)
----------------------

*	Added UDP network console support. This adds a 'netcon' command and
	associated variables, and a new boot menu item "Network Shell".

*	Added "ncon" to tools directory. Very simple net console client.

*	PCI serial support switch moved to NV flag 7.

CoLo 1.17 (2005-02-20)
----------------------

*	DHCP client now broadcasts DHCPREQUEST packets. It also includes the
	"server identifier" option from the server in the request. Fixed bug where
	we weren't appending an "end options" tag to the options. [RFC2131]

*	Dropped "no matching socket" diagnostic message.

*	Simplified display of memory size on boot.

*	Added "Cancel" option to 'paneld' menus.

*	Added horizontal scrolling menus to 'paneld' (Controlled by the '-h'
	option).

*	Added horizontal scrolling menus to shell. Controlled by an NV flag (see
	'nvflags' command in README.shell).

*	Reset menu timeout when a button is pressed.

*	Removed ARP diagnostic messages.

CoLo 1.16 (2004-12-16)
----------------------

*	Force menu choice of console on/off on upgrade from earlier versions.

*	Fix 'flash-tool' and 'flash' command to work with TMS29F040 Flash.

CoLo 1.15 (2004-11-28)
----------------------

*	Added link detection to network driver.

*	Added support for 21041 network controller (2700WG units).

*	Added clock speed detection (stage2 only) to ensure timeouts and delays are
	consistent regardless of CPU speed.

*	Added new command 'sleep'.

CoLo 1.14 (2004-10-26)
----------------------

*	Cleaned up build system and added cross compile support.
	[cross compile suggestion / gcc-3.4 link fix - Sebastien Vajda]
	[compiler flag suggestions - Thiemo Seufer]

*	Added new example boot script 'menu.colo' showing new features.

*	Variable expansion in the shell is now carried out before argument
	splitting.  This means variables can now span arguments e.g.

		var message "'first' 'second'"
		lcd {message}

*	Reworked the LCD menu command. The 'menu' command has been removed and
	replaced with a new command, 'select'. The new command justs sets a
	variable to the index of the selected item which can then be used with
	'goto'.

*	Added 'noop' command, which does nothing. When combined with the 'onfail'
	command it can be used for checking the existence of variables.

*	Added script commands 'goto', 'onfail', 'exit' and 'abort'.

*	Added positional marks in scripts using the '@' character.

*	Scripts can now be nested.

*	The 'console-speed' variable is now only available when the serial port is
	enabled.

*	Fixed up support for booting 64-bit kernels (32-bit support was reorganised
	too). 32-bit kernels must be linked for KSEG0 or KSEG1, 64-bit kernels for
	CKSEG0, CKSEG1 or XKPHYS.

*	Ensure we read from Galileo PCI RETRY/TIMEOUT register before writing to
	it. A comment in the kernel source indicates this might be necessary to
	stop Galileo locking up on early units.

*	Added a '-n' command line option to 'putlcd' which allows the user to write
	to a single line of the LCD without affecting the other line.

*	Added big fat warning to 'flash-tool' (just in case).

CoLo 1.13 (2004-08-30)
----------------------

*	Speeded up raw LCD access in LCD tools library.

*	Added 'on' and 'off' options to the 'serial' command.

*	Added 'colo-perm' and 'md5rom' to build.

*	Added missing include to paneld.c.

CoLo 1.12 (2004-08-20)
----------------------

*	The LCD tools 'putlcd', 'paneld' and 'e2fsck-lcd' are now built around a
	common library. The library has support for using direct access to the LCD
	via /dev/mem (the default) or driver access via the Cobalt LCD driver.

	To make the library use the Cobalt LCD driver you need to export an
	environment variable LCD_DEVICE with the name of the device. If LCD_DEVICE
	is an empty string the library will use the default device name, /dev/lcd.

*	Added a '-d' command line option to 'paneld' which makes it fork itself
	into the background on startup.

*	Added 'menu' boot shell command for implementing LCD menus in scripts.
	This feature makes it possible to choose from a list of multiple
	kernels/configurations to boot (see example in README.shell).

*	Fixed a couple of off by one errors in the 'boot' command.

*	Added 'console disable' to the non-volatile flags. This defers initialising
	the serial port until the boot shell is entered and should make it possible
	to use CoLo on units which don't have a serial port. This flag will be set

tools/copy-rom/copy-rom.c
/*
 * (C) P.Horton 2004,2005,2006
 *
 * $Id: copy-rom.c 186 2006-01-17 23:03:58Z pdh $
 *
 * This code is covered by the GNU General Public License. For details see the file "COPYING".
 */

#include <unistd.h>
#include <stdio.h>
#include <stdint.h>
#include <fcntl.h>
#include <string.h>
#include <errno.h>
#include <sys/mman.h>

#include "lib.h"
#include "md5.h"

#define APP_NAME					"copy-rom"

#define ROM_PHYS_ADDR			0x1fc00000
#define ROM_PHYS_SIZE			(512 << 10)

static int usage(void)
{
	puts("usage: " APP_NAME " [ file ]");

	return 1;
}

int main(int argc, char *argv[])
{
	struct MD5Context ctx;
	uint8_t dig[16];
	unsigned indx;
	int fd, copy;
	void *rom;

	if(argc > 2)
		return usage();

	fd = open("/dev/mem", O_RDONLY | O_SYNC);
	if(fd == -1) {
		fprintf(stderr, APP_NAME ": failed to open /dev/mem (%s)\n", strerror(errno));
		return 1;
	}

	rom = mmap(NULL, ROM_PHYS_SIZE, PROT_READ, MAP_SHARED, fd, ROM_PHYS_ADDR);
	if(rom == MAP_FAILED) {
		fprintf(stderr, APP_NAME ": failed to map /dev/mem (%s)\n", strerror(errno));
		return 1;
	}

	close(fd);

	if(argc > 1) {

		fd = creat(argv[1], 0664);
		if(fd == -1) {
			fprintf(stderr, APP_NAME ": failed to create file (%s)\n", strerror(errno));
			return 1;
		}

		for(indx = 0; indx < ROM_PHYS_SIZE;) {

			copy = write(fd, rom + indx, ROM_PHYS_SIZE - indx);
			if(copy == -1) {
				if(errno != EINTR) {
					fprintf(stderr, APP_NAME ": failed writing file (%s)\n", strerror(errno));
					return 1;
				}
			} else
				indx += copy;
		}

		close(fd);
	}

	MD5Init(&ctx);
	MD5Update(&ctx, rom, ROM_PHYS_SIZE);
	MD5Final(dig, &ctx);

	for(indx = 0; indx < sizeof(dig); ++indx)
		printf("%02x", dig[indx]);
	putchar('\n');

	return 0;
}

/* vi:set ts=3 sw=3 cin: */
